text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+It helps to think about the macro system as an extension to Javascript itself that we allow in v2 packages. Because the macros are allowed to appear in any published V2 package, and because the macros are _not_ a dependency that each package can control (you don't get to bring your own separate macro system version with you), it's important that we design a small core that we can support for the long-term. We probably can't make breaking changes to the macro system, we can only introduce new macros.
+
+(I'm proposing the macros live under `@ember/macros`. The current implementation of them lives in `@embroider/macros`.)
+
+The Javascript macros are:
+
+- importSync
+- getOwnConfig
+- getConfig
+- macroCondition
+- each
+- moduleExists
+- dependencySatisfies
+- failBuild
+
+The Handlebars macros are:
+
+- macroGetOwnConfig
+- macroGetConfig
+- macroCondition
+- macroDependencySatisfies
+- macroMaybeAttrs
+- macroFailBuild
+
+The difference in naming is because the JS macros get explicitly imported from `@ember/macros`, whereas the Handlebars macros do not, so they need an appropriate namespace prefix. (If we land template imports, I'm fine with adjusting this RFC to make the names align.)
+
+### JavaScript Macro: importSync
+
+```js
+import { importSync } from "@ember/macros";
+importSync("some-dependency").default;
+```
+
+`importSync` exists to do a thing that standard Javascript does not do: synchronous dynamic import. Ember historically needs synchronous dynamic import (it's what our runtime AMD `require` does). Until some future date at which Ember has migrated away from synchronous module resolution we need `importSync`.
+
+`importSync` is defined as behaving exactly like the standards-compliant `import()` except instead of returning `Promise<Module>` it returns `Module`.
+
+In this RFC we don't state explicitly what `importSync` compiles to. It compiles to whatever the Javascript bundler we're using supports in order to achieve synchronous dynamic import. For example, if we're internally using Webpack we can compile `importSync("something")` to `require("something")`, because Webpack supports CommonJS `require` anywhere.
+
+### JavaScript Macro: getOwnConfig
+
+```js
+// this example:
+import { getOwnConfig } from "@ember/macros";
+const shouldEnableCoolFeature = getOwnConfig().coolFeature;
+// might compile to:
+const shouldEnableCoolFeature = true;
+// assuming your ownConfig is `{ coolFeature: true }`.
+```
+
+`getOwnConfig()` behaves like a function that returns your `OwnConfig`, as determined by the return value of your `configure` **Build Hook**. You're allowed to chain property accesses (including array indices) off of `getOwnConfig()`. Since the `OwnConfig` is required to be JSON-serializable, any subset of it can be accessed this way, and we inline that value directly into the code.
```

Comment:
Is there reliable constant folding in babel and is that enabled by default? For example, would something like this work?

```js
const isCoolFeatureEnabled = "coolFeature" in getOwnConfig() && getOwnConfig().coolFeature.enabled === true;
const isCoolFeatureSupported = dependencySatisfies("ember-source", "^3.0.0");

if (macroCondition(isCoolFeatureEnabled && isCoolFeatureSupported)) {
  // ...
} else if (macroCondition(isCoolFeatureEnabled && !isCoolFeatureSupported)) {
  failBuild(...);
}
```

Reply:
Per the `macroCondition` rationale, isn't it better to do something like `getOwnConfig("some.nested.path")`, where it is clear what is static (whatever the path micro syntax allows you to express) and what is not (using the return value in JavaScript), rather than accidentally turning this into runtime code? (e.g. `getOwnConfig()["foo"]` vs `getOwnConfig()[foo]`)

Reply:
If the key is not found it's an error. If you want to check for an optional key, we explicitly support optional chaining: `getConfig().thing?.enabled`.

We don't currently support general constant folding, though implementing increasingly more analysis seems fine, as long as we only increase the set of cases that is supported and don't decrease it. There is no risk of accidentally allowing dynamic cases through because the macros will error if they can't statically understand a scenario.

> Per the macroCondition rationale, isn't it better to do something like getOwnConfig("some.nested.path"), where it is clear what is static (whatever the path micro syntax allows you to express) and what is not (using the return value in JavaScript), rather than accidentally turning this into runtime code? (e.g. getOwnConfig()["foo"] vs getOwnConfig()[foo])

That's not a bad idea. The cost of it is inventing a microsyntax. The benefit is that you can immediately see which parts are build-time evaluated.

I'm not actually sure if it's important to make those cases look different, however. Certainly `getOwnConfig()["foo"]` and `getOwnConfig()[foo]` are both legitimate cases, and they both work correctly today. They don't seem to have any ambiguity, they do exactly what the javascript says they do. The mental model for users should be that `getOwnConfig()` returns an object, and you can use it in every way you use any other object. We just happen to optimize how much of the object actually gets inline into the code when we happen to notice that you're only using some of it. But that is purely optimization, and doesn't impact correctness.

Reply:
> But that is purely optimization, and doesn't impact correctness.

Sure, but if you accidentally inlined a huge JSON blob that cannot be minified out, that could be about as bad as not eliminating a branch, which was the rationale of `macroCondition`. The config for things like eslint, etc, can be very big. Or it maybe all the localization strings for your entire app.

Reply:
An alternative to the micro syntax is something more like the design for `macroCondition`:

```js
import { ownConfig } from '@ember/macros';
````

Plus:

```
macroValue(ownConfig.foo.bar)
```

Verses:

```js
macroValue(ownConfig.foo)[this.dynamicConfigKey]
```

`macroValue` can ensure whatever is in the parens are inlined as a static literal value, otherwise it errors. If any part of the build wants to use `ownConfig` outside of `macroValue`, it can be emitted into a JavaScript module and the import rewritten to point to that generated module. That way, at least it's still only once per app at most, and none at all if all usages is wrapped in `macroValue()`.

Reply:
The import design here is massively preferable over the string-key/micro syntax style from the perspective of TypeScript consumers: microsyntaxes are extremely hostile to providing types for correctly, and these kinds of boundaries/leaves are often some of the first places teams can and should introduce TS in their codebase because their impact tends to be so distributed. (While we don’t officially support TS at the framework level, keeping these considerations in view will make TS users’ and the Typed Ember team’s lives much easier.)

Reply:
@chriskrycho that is a good point.


That said, I don't actually know how we could type `getOwnConfig` correctly anyway, because its meaning changes depending who's asking.

Reply:
`importSync` is also not possible to type automatically today, although the TS team has expressed interest in a feature that would make it possible. Today you need to say:

```importSync("./some/thing") as import("./some/thing")```

Reply:
I'll reread on Monday and see (and I may ping you in Discord to follow up on that), but I _think_ that there are a couple ways we could wrangle that, including with interface declaration merging. People still might need to do some small hoop-jumping-through to make it work, but the key is *small* (vs. the large hoops that you have to jump through for microsyntaxes).

Reply:
The not-yet-real feature would be something like:

```ts
importSync<T extends string>(specifier: T): import(T)
```